### ISSUE TITLE
Model is not getting saved after fine-tuning with weights and biases config: wandb_log_model

### ISSUE THREAD

**Issue by HeenaRajan** [axolotl-ai-cloud/axolotl#2337]
### Please check that this issue hasn't been reported before.

- [x] I searched previous [Bug Reports](https://github.com/axolotl-ai-cloud/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

With _wandb_log_model_  config set as either checkpoint or end, the model should get saved in Weights and Biases.

### Current behaviour

When setting this config as **checkpoint**, the model gets saved at each checkpoint but once the training ends it throws error: 
_You cannot perform fine-tuning on purely quantized models. Please attach trainable adapters on top of
 the quantized model to correctly perform fine-tuning. Please see: https://huggingface.co/docs/transformers/peft"
 for more details_

When setting this config **end**,  it throws the same error: 
<img width="934" alt="Image" src="https://github.com/user-attachments/assets/ccf8b980-8664-494d-900c-4b9b8d79549e" />

### Steps to reproduce

This is happening because in **Transformers(4.48.3**), in wandb integration -> WandbCallback: on_train_end function gets called.
https://github.com/huggingface/transformers/blob/main/src/transformers/integrations/integration_utils.py

and then Transformers-> trainer.py Line 557, there are conditions to check if it is a peft model, quantized base model and so
https://github.com/huggingface/transformers/blob/main/src/transformers/trainer.py

I noticed that at each checkpoints, **the model type is Peft model and on train end it is FullyShardedDataParallel model**
since I am using FSDP config for multi-GPU training.

### Config yaml

```yaml
base_model: mistralai/Mistral-7B-Instruct-v0.2
model_type: AutoModelForCausalLM
tokenizer_type: LlamaTokenizer
trust_remote_code: true

is_mistral_derived_model: true

bnb_config_kwargs:
  bnb_4bit_quant_type: nf4
  bnb_4bit_use_double_quant: false
  bnb_4bit_quant_storage: float16

load_in_8bit: false
load_in_4bit: true

bf16: false
fp16: true
tf32: false

datasets:
  - path: abc/ICSR_classification
    type: alpaca
    train_on_split: train
    
val_set_size: 0.01
    
dataset_prepared_path: last_run_prepared

hub_model_id: abc/ICSR_classification
push_to_hub: True
hub_strategy: end
hf_use_auth_token: true

sequence_len: 700
sample_packing: false
pad_to_sequence_len: true

adapter: qlora

# LoRA hyperparameters
lora_r: 256
lora_alpha: 128
lora_dropout: 0.05
lora_target_linear: true
lora_fan_in_fan_out: false
lora_modules_to_save:

# Where to save the finetuned model to
output_dir: ./outputs/qlora_out

# wandb configuration if you're using it
wandb_mode: online
wandb_project: finetuning_ICSR_classification
wandb_entity: abc
wandb_name: experimental_run
# wandb_run_id: 
wandb_log_model: end

# Training hyperparameters
gradient_accumulation_steps: 4
micro_batch_size: 8
eval_batch_size: 8
num_epochs: 1
warmup_ratio: 0.03
learning_rate: 0.0002
optimizer: paged_adamw_32bit
lr_scheduler: cosine

eval_steps: 50
eval_strategy:
save_strategy: steps
save_steps: 50
logging_steps: 50

train_on_inputs: false
group_by_length: false

# gradient_checkpointing: false
# gradient_checkpointing_kwargs:
#   use_reentrant: true

early_stopping_patience:
resume_from_checkpoint:

xformers_attention:
flash_attention: false

loss_watchdog_threshold: 5.0
loss_watchdog_patience: 3

weight_decay: 0.0
max_grad_norm: 0.3

fsdp:
  - full_shard
  - auto_wrap
fsdp_config:
  fsdp_limit_all_gathers: true
  fsdp_sync_module_states: true
  fsdp_offload_params: false
  fsdp_use_orig_params: false
  fsdp_cpu_ram_efficient_loading: true
  fsdp_activation_checkpointing: true
  fsdp_transformer_layer_cls_to_wrap: MistralDecoderLayer
  fsdp_state_dict_type: FULL_STATE_DICT
  fsdp_auto_wrap_policy: TRANSFORMER_BASED_WRAP
special_tokens:

strict: false
```

### Which Operating Systems are you using?

- [x] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.11

### axolotl branch-commit

main

### Acknowledgements

- [x] My issue title is concise, descriptive, and in title casing.
- [x] I have searched the existing issues to make sure this bug has not been reported yet.
- [x] I am using the latest version of axolotl.
- [x] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by NanoCode012** [comment]
Hey, thanks for report.

> When setting this config as checkpoint, the model gets saved at each checkpoint but once the training ends it throws error:

Does it upload existing checkpoints to wandb? It only crashes on train_end correct? How about last checkpoint before train end?

We don't really test wandb log model, so this issue might've slipped through.

**Comment by HeenaRajan** [comment]
Hi @NanoCode012 
Yes, exactly. the existing checkpoints are uploaded on wandb, the last checkpoint before train end ad well. 
It crashes when training ends and  WandbCallback: on_train_end function is called.

**Comment by NanoCode012** [comment]
That's weird. I wonder if this makes more sense to be reported upstream in transformers regarding wandb integration? All we do is pass wandb configs along.

**Comment by NanoCode012** [comment]
Hey @HeenaRajan , just checking back in. Did you raise an issue about this upstream?
